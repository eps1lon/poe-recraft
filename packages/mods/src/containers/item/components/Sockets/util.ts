import Item from '../../Item';
import { Socket, SocketColor } from './types';
import ValueRange from '../../../../calculator/ValueRange';

/**
 * creates a Socket according to the highest attr requirement
 * if two or more requirements are equal the color is chosen
 * with the priority Green > Red > Blue
 * @param item
 */
export function rollRequirementBiasedStable(item: Item): Socket {
  const { dex, int, str } = item.requirements.list();
  const color_by_attr = [
    { value: dex.value, color: SocketColor.green },
    { value: str.value, color: SocketColor.red },
    { value: int.value, color: SocketColor.blue },
    // descending
  ].sort((a, b) => -ValueRange.cmpMax(a.value, b.value));

  return {
    color: color_by_attr[0].color,
  };
}
