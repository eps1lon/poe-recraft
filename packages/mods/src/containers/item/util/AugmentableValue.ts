/**
 * a numeric value that was augmented in some way
 *
 * {augmented} is true if the the base value was changed to {value}
 */
export default interface AugmentableValue {
  value: number | [number, number];
  augmented: boolean;
}

export function isZero(
  augmentable: { value: AugmentableValue['value'] } | null | undefined,
): boolean {
  if (augmentable == null) {
    return true;
  }
  const { value } = augmentable;

  return (
    value === 0 || (Array.isArray(value) && value[0] === 0 && value[1] === 0)
  );
}

export function anyCmp(
  augmentable: { value: AugmentableValue['value'] },
  cmp: (n: number) => boolean,
): boolean {
  const { value } = augmentable;

  if (typeof value === 'number') {
    return cmp(value);
  } else {
    return value.some(cmp);
  }
}
