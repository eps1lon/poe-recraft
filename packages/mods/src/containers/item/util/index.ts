export { default as AugmentableValue } from './AugmentableValue';
