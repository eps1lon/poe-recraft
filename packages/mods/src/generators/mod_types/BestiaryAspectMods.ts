import { GeneratorDetails } from '../Generator';
import ViewOnlyOrb from '../ViewOnlyOrb';
import Item from '../../containers/item';
import { Mod } from '../../mods';
import { ModProps } from '../../schema';

/**
 * Generator for all mods that can be crafted at the Blood Altar with Spirit Beasts
 */
export default class BestiaryAspectMods extends ViewOnlyOrb {
  /**
   * list of mod ids that can be crafted with Spirit Beasts
   */
  public static readonly ASPECT_MODS: Set<Mod['props']['id']> = new Set([
    'GrantsCatAspectCrafted',
    'GrantsBirdAspectCrafted',
    'GrantsSpiderAspectCrafted',
    'GrantsCrabAspectCrafted',
  ]);

  public static build(mods: ModProps[]): BestiaryAspectMods {
    return new BestiaryAspectMods(
      mods
        .filter(BestiaryAspectMods.isBestiaryAspect)
        .map(props => new Mod(props)),
    );
  }

  /**
   *
   * @param mod
   * @returns true if the mod can only crafted with the Blood Altar from a
   *          Spirit Beast
   */
  public static isBestiaryAspect(mod: ModProps): boolean {
    return BestiaryAspectMods.ASPECT_MODS.has(mod.id);
  }

  public modsFor(
    item: Item,
    whitelist: string[] = [],
  ): Array<GeneratorDetails<Mod>> {
    // omit everything spawnable related. crafting is deterministic
    return super
      .modsFor(item, whitelist.concat('no_matching_tags', 'spawnweight_zero'))
      .map(({ spawnable, spawnweight, ...details }) => details);
  }
}
