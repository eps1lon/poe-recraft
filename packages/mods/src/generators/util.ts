import { GeneratorDetails } from './Generator';

// filter nothing
const allowAll = () => true;

export interface WithSpawnchance {
  spawnchance: number;
}

export const calculateSpawnchance = (
  collection: Array<GeneratorDetails<any>>,
  filter: (details: GeneratorDetails<any>) => boolean = allowAll,
): Array<GeneratorDetails<any> & WithSpawnchance> => {
  const sum_spawnweight = collection.reduce((sum, details) => {
    if (details.spawnweight != null && filter(details)) {
      return sum + details.spawnweight;
    } else {
      return sum;
    }
  }, 0);

  return collection.map(details => {
    let spawnchance = 0;

    if (details.spawnweight != null && filter(details)) {
      spawnchance = details.spawnweight / sum_spawnweight;
    }

    return {
      ...details,
      spawnchance,
    };
  });
};
