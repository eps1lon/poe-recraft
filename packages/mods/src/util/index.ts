export { Flags, anySet } from './Flags';
export { default as MetaData } from './MetaData';
