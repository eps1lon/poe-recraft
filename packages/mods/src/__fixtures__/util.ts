import * as fs from 'fs';
import * as path from 'path';

import {
  AtlasNodeProps,
  BaseItemTypeProps,
  CraftingBenchOptionsProps,
  EssenceProps,
  ModProps,
} from '../schema';

import AtlasNode from '../containers/AtlasNode';
import Item from '../containers/item/Item';
import { Essence, MasterBenchOption } from '../generators';
import Mod from '../mods/Mod';
import PropsTable, { TableProps } from '../helpers/PropsTable';
import { Buildable } from '../interfaces/Buildable';

const json_cache: Partial<Tables> = {};

export const createTable = <P extends TableProps, T, A1>(
  file: keyof Tables,
  constructor: Buildable<P, T, A1>,
  arg1: A1,
) => {
  if (json_cache[file] == null) {
    const body = fs.readFileSync(
      path.join(__dirname, `../../data/${file}.json`),
    );

    const json = JSON.parse(body.toString());

    // @ts-ignore
    json_cache[file] = new PropsTable(json, constructor, arg1);
  }

  return json_cache[file];
};

interface Tables {
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>;
  craftingbenchoptions: PropsTable<
    CraftingBenchOptionsProps,
    MasterBenchOption,
    undefined
  >;
  essences: PropsTable<EssenceProps, Essence, ModProps>;
  items: PropsTable<BaseItemTypeProps, Item, undefined>;
  mods: PropsTable<ModProps, Mod, undefined>;
}

export const createTables = () => {
  const tables = {
    atlas: createTable('atlas', AtlasNode, undefined),
    craftingbenchoptions: createTable(
      'craftingbenchoptions',
      MasterBenchOption,
      undefined,
    ),
    items: createTable('items', Item, undefined),
    mods: createTable('mods', Mod, undefined),
  } as Tables;

  // @ts-ignore: its not undefined though?!
  tables.essences = createTable('essences', Essence, tables.mods.all());

  return tables;
};

export const byWorldAreaId = (id: string) => (node: AtlasNodeProps) =>
  node.world_area.id === id;

export const fromWorldAreaId = (
  id: string,
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>,
): AtlasNode => {
  return atlas.from(node => node.world_area.id === id);
};

export const byEssenceName = (name: string) => (essence: EssenceProps) =>
  essence.base_item_type.name === name;
