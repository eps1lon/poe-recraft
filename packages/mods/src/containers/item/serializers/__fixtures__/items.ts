import { createTables } from '../../../../__fixtures__/util';

const { items, mods } = createTables();

export default [
  items
    .fromName('Platinum Kris')
    .rarity.set('magic')
    .addMod(
      mods.fromId('LocalIncreasedPhysicalDamagePercentAndAccuracyRating2'),
    )
    .setProperty('mirrored', true),
  items
    .fromName('Hubris Circlet')
    .rarity.set('rare')
    .asShaperItem()
    .addMod(mods.fromId('IncreasedLife0'))
    .addMod(mods.fromId('LocalIncreasedEnergyShield2'))
    .addMod(mods.fromId('LocalIncreasedEnergyShieldPercentAndStunRecovery3'))
    .addMod(mods.fromId('FireResist1'))
    .addMod(mods.fromId('ReducedLocalAttributeRequirements1'))
    .addMod(mods.fromId('AdditionalMinesPlacedSupportedUber1_'))
    .setProperty('corrupted', true),
];
