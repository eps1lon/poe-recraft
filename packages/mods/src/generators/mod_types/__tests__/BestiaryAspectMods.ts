import { createTables } from '../../../__fixtures__/util';
import BestiaryAspectMods from '../BestiaryAspectMods';

const { items, mods } = createTables();

it('ignores internal spawnweights', () => {
  const item = items.fromName('Iron Greaves');
  const generator = BestiaryAspectMods.build(mods.all());

  const bestiaryMods = generator.modsFor(item, ['domain_full']);

  expect(bestiaryMods.length).toBeGreaterThan(0);
  expect(
    bestiaryMods.map(({ mod, ...details }) => ({
      mod: mod.props.id,
      ...details,
    })),
  ).toMatchSnapshot();
});
