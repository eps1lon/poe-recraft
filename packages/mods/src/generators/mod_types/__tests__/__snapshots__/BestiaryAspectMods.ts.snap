// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ignores internal spawnweights 1`] = `
Array [
  Object {
    "applicable": Object {
      "above_lld_level": false,
      "already_present": false,
      "domain_full": true,
      "lower_ilvl": false,
      "wrong_domain": false,
    },
    "mod": "GrantsCatAspectCrafted",
  },
  Object {
    "applicable": Object {
      "above_lld_level": false,
      "already_present": false,
      "domain_full": true,
      "lower_ilvl": false,
      "wrong_domain": false,
    },
    "mod": "GrantsBirdAspectCrafted",
  },
  Object {
    "applicable": Object {
      "above_lld_level": false,
      "already_present": false,
      "domain_full": true,
      "lower_ilvl": false,
      "wrong_domain": false,
    },
    "mod": "GrantsSpiderAspectCrafted",
  },
  Object {
    "applicable": Object {
      "above_lld_level": false,
      "already_present": false,
      "domain_full": true,
      "lower_ilvl": false,
      "wrong_domain": false,
    },
    "mod": "GrantsCrabAspectCrafted",
  },
]
`;
